### PR TITLE
Add container app module

### PR DESCRIPTION
**Pull request by trapeznikov**
# [1301](https://github.com/aztfmod/terraform-azurerm-caf/issues/1301)

## PR Checklist

---

- [x] I have added example(s) inside the [./examples/] folder
- [x] I have added the example(s) to the integration test list for [normal (~30 minutes)](./workflows/standalone-scenarios.json) or [long runner >30 minutes](./workflows/standalone-scenarios-longrunners.json)
- [x] I have checked the [coding conventions as per the wiki](https://github.com/aztfmod/terraform-azurerm-caf/wiki)
- [x] I have checked to ensure there aren't other open Pull Requests for the same update/change?

## Does this introduce a breaking change

- [ ] YES
- [x] NO

### CODE REVIEW

**Review by arnaudlh — CHANGES_REQUESTED** [review]
Thanks for the PR @trapeznikov, Im running examples and getting:

```
│ Error: Reference to undeclared input variable
│ 
│   on module.tf line 106, in module "example":
│  106:     container_app                         = var.container_app
│ 
│ An input variable with the name "container_app" has not been declared. This variable can be declared with a variable
│ "container_app" {} block.
╵
╷
│ Error: Reference to undeclared input variable
│ 
│   on module.tf line 107, in module "example":
│  107:     container_app_dapr_component          = var.container_app_dapr_component
│ 
│ An input variable with the name "container_app_dapr_component" has not been declared. This variable can be declared with
│ a variable "container_app_dapr_component" {} block.
╵
╷
│ Error: Reference to undeclared input variable
│ 
│   on module.tf line 108, in module "example":
│  108:     container_app_environment             = var.container_app_environment
│ 
│ An input variable with the name "container_app_environment" has not been declared. This variable can be declared with a
│ variable "container_app_environment" {} block.
╵
╷
│ Error: Reference to undeclared input variable
│ 
│   on module.tf line 109, in module "example":
│  109:     container_app_environment_certificate = var.container_app_environment_certificate
│ 
│ An input variable with the name "container_app_environment_certificate" has not been declared. This variable can be
│ declared with a variable "container_app_environment_certificate" {} block.
╵
╷
│ Error: Reference to undeclared input variable
│ 
│   on module.tf line 110, in module "example":
│  110:     container_app_environment_storage     = var.container_app_environment_storage
│ 
│ An input variable with the name "container_app_environment_storage" has not been declared. This variable can be declared
│ with a variable "container_app_environment_storage" {} block.
╵
Terraform plan return code: 1
Error 1 on or near line 57: Error running terraform plan; exiting with status 1
```

**Comment by trapeznikov** [comment]
@arnaudlh need to add container_app to https://github.com/Azure/caf-terraform-landingzones/tree/main/caf_solution. What's the process of adding a new feature to these 2 repos? Should the change first go to `caf-terraform-landingzones`?

**Comment by arnaudlh** [comment]
> @arnaudlh need to add container_app to https://github.com/Azure/caf-terraform-landingzones/tree/main/caf_solution. What's the process of adding a new feature to these 2 repos? Should the change first go to `caf-terraform-landingzones`?

1. update the module with examples working, so people can leverage it as standalone
2. update the azure repo with the new var names and module version constraint.

**Comment by trapeznikov** [comment]
@arnaudlh fixed the variables file in the example module, it should run fine now. Sorry for the confusion, I was just testing it differently.

**Comment by mresetar-croz** [comment]
I am currently working with this PR code. I hope it gets merged soon. Maybe not the right place but please follow https://github.com/hashicorp/terraform-provider-azurerm/issues/21747 as workload profiles support is something in the future that should be supported by CAF as well.

File: examples/module.tf [hunk 1]
```diff
@@ -93,31 +93,36 @@ module "example" {
...
+    batch_applications                    = var.batch_applications
+    batch_certificates                    = var.batch_certificates
+    batch_jobs                            = var.batch_jobs
+    batch_pools                           = var.batch_pools
+    bastion_hosts                         = var.bastion_hosts
+    container_app                         = var.container_app
+    container_app_dapr_component          = var.container_app_dapr_component
+    container_app_environment             = var.container_app_environment
+    container_app_environment_certificate = var.container_app_environment_certificate
+    container_app_environment_storage     = var.container_app_environment_storage
```
Comment (jvdhaterdwestland) [inline examples/module.tf:110]: Currently all container_app related are in singular form, this should be plural.

File: locals.tf [hunk 2]
```diff
@@ -74,31 +74,36 @@ locals {
...
+    batch_applications                    = try(var.compute.batch_applications, {})
+    batch_certificates                    = try(var.compute.batch_certificates, {})
+    batch_jobs                            = try(var.compute.batch_jobs, {})
+    batch_pools                           = try(var.compute.batch_pools, {})
+    container_app                         = try(var.compute.container_app, {})
+    container_app_dapr_component          = try(var.compute.container_app_dapr_component, {})
+    container_app_environment             = try(var.compute.container_app_environment, {})
+    container_app_environment_certificate = try(var.compute.container_app_environment_certificate, {})
+    container_app_environment_storage     = try(var.compute.container_app_environment_storage, {})
+    container_groups                      = try(var.compute.container_groups, {})
```
Comment (jvdhaterd) [inline locals.tf:92]: Imho the container_app related should be in a plural form, matching the rest.

Comment (trapeznikov) [reply locals.tf:92]: Thanks for your feedback, I will fix it.

File: modules/compute/container_app/container_app.tf [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+resource "azurerm_container_app" "ca" {
+  name                         = azurecaf_name.ca.result
+  resource_group_name          = local.resource_group_name
+  container_app_environment_id = var.container_app_environment_id
+  revision_mode                = var.settings.revision_mode
+  tags                         = merge(local.tags, try(var.settings.tags, null))
+
+  template {
```
[the remark this answers is not in the document]
Comment (trapeznikov) [reply modules/compute/container_app/container_app.tf:18]: @chatelain-io added support for scale rules.

**Comment by trapeznikov** [comment]
@arnaudlh I resolved all the comment and rebased the branch. Can you please review?

**Review by jvdhaterd — APPROVED** [review]

**Comment by trapeznikov** [comment]
@arnaudlh @LaurentLesle can you please review this PR?

File: modules/compute/container_app_environment/container_app_environment.tf [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+resource "azurerm_container_app_environment" "cae" {
+  name                           = azurecaf_name.cae.result
+  location                       = local.location
+  resource_group_name            = local.resource_group_name
+  log_analytics_workspace_id     = can(var.settings.log_analytics_workspace_id) ? var.settings.log_analytics_workspace_id : var.diagnostics.log_analytics[var.settings.log_analytics_key].id
+  infrastructure_subnet_id       = try(var.subnet_id, null)
+  internal_load_balancer_enabled = try(var.settings.internal_load_balancer_enabled, false)
+  tags                           = merge(local.tags, try(var.settings.tags, null))
```
Comment (iriahk89) [inline modules/compute/container_app_environment/container_app_environment.tf:18]: please help to add support for another 2 parameter: dapr_application_insights_connection_string & zone_redundancy_enabled

**Review by iriahk89 — CHANGES_REQUESTED** [review]
@trapeznikov I reviewed the PR and would really appreciate if you could add those 2 features and resolve the minor bug as one of my customers also waiting for this resource. Let me know if you need helps. Thanks!

File: modules/compute/container_app_environment/container_app_environment.tf (hunk 4 above)
Comment (arnaudlh) [reply modules/compute/container_app_environment/container_app_environment.tf:18]: @trapeznikov good PR - thank you! if you dont mind adding the 2 attributes mentionned by @iriahk89, I think we are good to go!

**Comment by iriahk89** [comment]
@trapeznikov I reviewed the PR and would really appreciate if you could add those 2 features and resolve the minor bug as one of my customers also waiting for this resource. Let me know if you need helps. Thanks!

File: modules/compute/container_app_environment/container_app_environment.tf (hunk 4 above)
Comment (trapeznikov) [reply modules/compute/container_app_environment/container_app_environment.tf]: @arnaudlh @iriahk89 done. Please review.

**Review by iriahk89 — APPROVED** [review]
LGTM. we can add the integration with app insight combined object later for dapr_application_insights_connection_string if needed

**Review by arnaudlh — APPROVED** [review]
LGTM

**Comment by arnaudlh** [comment]
thanks for your contribution @trapeznikov   and for your reviews @iriahk89 @jvdhaterd

**Pull request merged**